pex/tools/commands/venv.py

```diff
@@ -0,0 +1,236 @@
+                if e.errno != errno.EEXIST:
+                    raise e
+
+        for f in files:
+            src_entry = os.path.join(root, f)
+            dst_entry = os.path.join(dst, os.path.relpath(src_entry, src))
+            try:
+                if link:
+                    try:
+                        os.link(src_entry, dst_entry)
+                        continue
+                    except OSError as e:
+                        if e.errno != errno.EXDEV:
+                            raise e
+                        link = False
+                shutil.copy(src_entry, dst_entry)
+            except OSError as e:
+                if e.errno == errno.EEXIST:
+                    pex_warnings.warn(
+                        "Failed to overwrite {} with {}: {}".format(dst_entry, src_entry, e)
+                    )
+                    if not collisions_ok:
+                        raise e
+
+
+class Venv(Command):
+    """Creates a venv from the PEX file."""
+
+    def add_arguments(self, parser):
+        # type: (ArgumentParser) -> None
+        parser.add_argument(
+            "venv",
+            nargs=1,
+            metavar="PATH",
+            help="The directory to create the virtual environment in.",
+        )
+        parser.add_argument(
+            "-b",
+            "--bin-path",
+            choices=("prepend", "append"),
+            default=None,
+            help="Add the venv bin dir to the PATH in the __main__.py script.",
+        )
+        parser.add_argument(
+            "-f",
+            "--force",
+            action="store_true",
+            default=False,
+            help="If the venv directory already exists, overwrite it.",
+        )
+        parser.add_argument(
+            "--collisions-ok",
+            action="store_true",
+            default=False,
+            help=(
+                "Don't error if population of the venv encounters distributions in the PEX file "
+                "with colliding files."
+            ),
+        )
+        parser.add_argument(
+            "-p",
+            "--pip",
+            action="store_true",
+            default=False,
+            help="Add pip to the venv.",
+        )
+
+    def run(
+        self,
+        pex,  # type: PEX
+        options,  # type: Namespace
+    ):
+        # type: (...) -> Result
+
+        venv = Virtualenv.create(options.venv[0], interpreter=pex.interpreter, force=options.force)
+
+        # 1. Populate the venv with the PEX contents.
+        pex_info = pex.pex_info()
+        if zipfile.is_zipfile(pex.path()):
+            PEXEnvironment.explode_code(
+                pex.path(), pex_info, venv.site_packages_dir, exclude=("__main__.py",)
+            )
+        else:
+            _copytree(
+                src=pex.path(),
+                dst=venv.site_packages_dir,
+                exclude=(pex_info.internal_cache, pex_builder.BOOTSTRAP_DIR, "__main__.py"),
+            )
+
+        for dist in pex.activate():
```

Comment:
I had thought of "activate" as meaning "add to the pythonpath of the current process", so it's interesting to see it used like this. Would splitting that method into "iterate over what would have been activated" vs "activate" make sense?

Reply:
That might make sense although it is neither here nor there for tool use where the side-effect of the distributions getting added to the sys.path here has no relevant effect:
https://github.com/pantsbuild/pex/blob/12ad6a63f988c07065cebbfb6aa4102a6d3fa094/pex/environment.py#L476

Reply:
This is a good deal thornier than it may look. The code paths involved are relevant in the resolve-from-pex work though so I'll defer to that work.